docs/design/operator.md

```diff
@@ -0,0 +1,407 @@
+# Ceph CSI Operator Design Document
+
+## Introduction
+
+This document outlines the design of the Ceph-CSI Operator, which aims to
+provide Kubernetes native management interfaces for Ceph-CSI drivers (CephFS,
+RBD, and NFS) for Kubernetes based environments. The operator automates the
+deployment, configuration, and management of these drivers using new Kubernetes
+APIs defined as a set of Custom Resource Definitions (CRDs).
+
+## System Overview
+
+The Ceph-CSI Operator is a Kubernetes operator that offers a Kubernetes native
+management interfaces for Ceph CSI drivers. It provides a set of new Kubernetes
+APIs (CRDs) for governing deployment and management of Ceph-CSI drivers, as
+well as connection details to Ceph clusters. The operator ensures that the
+desired state specified in the CRDs is maintained in the cluster.
+
+## Assumptions and Dependencies
+
+Kubernetes cluster is available and operational. Ceph cluster is set up and
+accessible from the Kubernetes cluster. Users are familiar with Kubernetes and
+Ceph concepts. CRD and operator versions are compatible with the Kubernetes
+version in use.
+
+## General Constraints
+
+Resource limitations of the Kubernetes cluster (CPU, memory). Network latency
+and connectivity between the Kubernetes cluster and Ceph cluster. Watching for
+changes on the ceph cluster and updating internal details.  Security policies
+and RBAC configurations in the Kubernetes cluster.
+
+## Goals and Guidelines
+
+- Automate the deployment and lifecycle for Ceph-CSI drivers.
+- Provide a consistent and flexible configuration mechanism using CRDs.
+- Enable namespace-scoped configurations for different Ceph-CSI drivers.
+- Ensure high availability and scalability of the operator.
+- Support multiple deployment methods for the operator (single-namespace,
+  multi-namespace, all-namespaces).
+- Support multiple operator deployment in a single cluster when running in a
+  single-namespace mode.
+
+## Development Methods
+
+Use operator framework which is higly used in operator developed and also by
+following Kubernetes best practices for operator development.
+
+## Architecture
+
+```mermaid
+graph TD
+    A[CRD Changes] --> B[Operator]
+    B -->  D[Configure CephFS CSI Driver]
+    B -->  E[Configure RBD CSI Driver]
+    B -->  F[Configure NFS CSI Driver]
+
+```
+
+In this updated diagram:
+
+- **CRD Changes**: Represents changes made to Custom Resource Definitions,
+  which trigger actions in the operator.
+- **Operator**: Listens for changes in CRDs and initiates the installation and
+  configuration of CSI drivers.
+- **Configure CephFS, NFS, RBD**: Actions performed by the operator to install
+  and configure the respective CSI drivers based on the CRD changes.
```

Comment:
This architecture is one-way only. `CRD changes` -> `Configure drivers`. I know that CSI drivers are usually pretty simple and can usually just be a deployment and daemonset, so maybe this really is the architecture. But the operator pattern follows control theory, where system states also result in the controller taking actions. 

Are there any other :
- are there ConfigMaps that cause reconciliations?
- does the controller respond to any other events like nodes going online/offline, or getting new labels applied?


I feel like this is a bigger question: is this operator taking on the node loss fencing behavior? https://github.com/rook/rook/blob/682d639532534068a6154664119e10f0016f547d/design/ceph/node-loss-rbd-cephfs.md

I feel that, since the node fencing is both Kube-specific and involves CSI-Addons CRDs, it makes sense that the Ceph-CSI driver would be the right controller for this, so that non-Rook users could benefit from the fencing that RBD and CephFS need to function well in failure conditions.

Reply:
>are there ConfigMaps that cause reconciliations?

There will be 2 controllers in cephcsi-operator
* One which takes care of deployment
* The one Reconciles all the configuration and update the configmap which is already mounted to the csi pods.

Good thought but I haven't thought about making csi-addons activities as part of the core operator we can decide on that where the operator needs to create extra CR but we can take this in the next iteration not in the first phase.

Reply:
Resolving as new issue is opened for it.